Docs: Keyring Migration & Events (#8895)
cosmos/cosmos-sdk: docs/core/encoding.md

### Gogoproto

> Modules are encouraged to utilize Protobuf encoding for their respective types. In the SDK, we use a specific implementation of the Protobuf spec, called [Gogoproto](https://github.com/gogo/protobuf), which offers speed and DX improvements compared to the official [Google implementation](https://github.com/protocolbuffers/protobuf).

Reviewer: ```suggestion
Modules are encouraged to utilize Protobuf encoding for their respective types. In the SDK, we use the [Gogoproto](https://github.com/gogo/protobuf) specific implementation of the Protobuf spec that offers speed and DX improvements compared to the official [Google protobuf implementation](https://github.com/protocolbuffers/protobuf).
```

### Guidelines for protobuf message definitions

> In addition to [following official guidelines](https://developers.google.com/protocol-buffers/docs/proto3#simple), we recommend to use these annotations in .proto files when dealing with interfaces:

Reviewer: ```suggestion
In addition to [following official Protocol Buffer guidelines](https://developers.google.com/protocol-buffers/docs/proto3#simple), we recommend using these annotations in .proto files when dealing with interfaces:
```

> - fields which accept interfaces should be annotated with `cosmos_proto.accepts_interface`
>   using the same full-qualified name passed as `protoName` to `InterfaceRegistry.RegisterInterface`
> - interface implementations should be annotated with `cosmos_proto.implements_interface`
>   using the same full-qualified name passed as `protoName` to `InterfaceRegistry.RegisterInterface`

Reviewer: ```suggestion
- use `cosmos_proto.accepts_interface` to annote fields that accept interfaces
- pass the same fully qualified name as `protoName` to `InterfaceRegistry.RegisterInterface`
- annotate interface implementations with `cosmos_proto.implements_interface`
- pass the same fully qualified name as `protoName` to `InterfaceRegistry.RegisterInterface`
```

Reviewer: Suggest that we move the action to the front of each line item "use, annotate, pass" so that the action is clear and list items have parallel construction

### Transaction Encoding

> Another important use of Protobuf is the encoding and decoding of
> [transactions](./transactions.md). Transactions are defined by the application or
> the SDK, but passed to the underlying consensus engine in order to be relayed to
> other peers. Since the underlying consensus engine is agnostic to the application,
> it only accepts transactions in the form of raw bytes. The encoding is done by an
> object called `TxEncoder` and the decoding by an object called `TxDecoder`.

Reviewer: ```suggestion
Another important use of Protobuf is the encoding and decoding of
[transactions](./transactions.md). Transactions are defined by the application or
the SDK but are then passed to the underlying consensus engine to be relayed to
other peers. Since the underlying consensus engine is agnostic to the application,
the consensus engine accepts only transactions in the form of raw bytes. 

- The `TxEncoder` object performs the encoding.
- The `TxDecoder` object performs the decoding.
```

+++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0-rc4/types/tx_msg.go#L83-L87

A standard implementation of both these objects can be found in the [`auth` module](../../x/auth/spec/README.md):

+++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0-rc4/x/auth/tx/decoder.go

+++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0-rc4/x/auth/tx/encoder.go

> Please also refer to [ADR-020](../architecture/adr-020-protobuf-transaction-encoding.md) on the details of how a transaction is encoded.

Reviewer: ```suggestion
See [ADR-020](../architecture/adr-020-protobuf-transaction-encoding.md) for details of how a transaction is encoded.
```

> ### Interface Encoding and Usage of `Any`s
> 
> The Protobuf DSL is strongly typed, which can make inserting variable-typed fields difficult. Imagine we want to create a `Profile` protobuf message, which serves as a wrapper over [an account](../basics/accounts.md):

Reviewer: ```suggestion
### Interface Encoding and Usage of `Any`

The Protobuf DSL is strongly typed, which can make inserting variable-typed fields difficult. Imagine we want to create a `Profile` protobuf message that serves as a wrapper over [an account](../basics/accounts.md):
```

Reviewer: we can usually use singular, so if we have tick marks around the subject we can omit the s for plural

```proto
message Profile {
  // account is the account associated to a profile.
  cosmos.auth.v1beta1.BaseAccount account = 1;
  // bio is a short description of the account.
  string bio = 4;
}
```

> In the above `Profile`, we hardcoded `account` as a `BaseAccount`. However, there are several other types of [user accounts related to vesting](../../x/auth/spec/05_vesting.md), such as `BaseVestingAccount` or `ContinuousVestingAccount`. All of these accounts are different, but they all implement the `AccountI` interface. How would one create a `Profile` which allows all these types of accounts, with an `account` field accepting an `AccountI` interface?

Reviewer: ```suggestion
In this `Profile` example, we hardcoded `account` as a `BaseAccount`. However, there are several other types of [user accounts related to vesting](../../x/auth/spec/05_vesting.md), such as `BaseVestingAccount` or `ContinuousVestingAccount`. All of these accounts are different, but they all implement the `AccountI` interface. How would you create a `Profile` that allows all these types of accounts with an `account` field that accepts an `AccountI` interface?
```

Reviewer: tip: the spatial references are covered by many accessibility rules in the dev doc universe, a good recommendation is described in the [Microsoft Style Guide](https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/above)

Reviewer: oooh, we can usually use second person "you" instead of "one" or other references

+++ https://github.com/cosmos/cosmos-sdk/blob/v0.42.1/x/auth/types/account.go#L307-L330

> In [ADR-019](../architecture/adr-019-protobuf-state-encoding.md), it has been decided to use [`Any`](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto)s to encode interfaces in protobuf. An `Any` contains an arbitrary serialized message as bytes, along with a URL that acts as a globally unique identifier for and resolves to that message's type. It allows us to pack arbitrary Go types inside protobuf messages. Our new `Profile` then looks like:

Reviewer: ```suggestion
In [ADR-019](../architecture/adr-019-protobuf-state-encoding.md), it has been decided to use [`Any`](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto)s to encode interfaces in protobuf. An `Any` contains an arbitrary serialized message as bytes, along with a URL that acts as a globally unique identifier for and resolves to that message's type. This strategy allows us to pack arbitrary Go types inside protobuf messages. Our new `Profile` then looks like:
```

Reviewer: I think we mean "This strategy" where we have "it"

```protobuf
message Profile {
  // account is the account associated to a profile.
  google.protobuf.Any account = 1 [
    (cosmos_proto.accepts_interface) = "AccountI"; // Asserts that this field only accepts Go types implementing `AccountI`. It is purely informational for now.
  ];
  // bio is a short description of the account.
  string bio = 4;
}
```

To add an account inside a profile, we need to "pack" it inside an `Any` first, using `codectypes.NewAnyWithValue`:

```go
var myAccount AccountI
myAccount = ... // Can be a BaseAccount, a ContinuousVestingAccount or any struct implementing `AccountI`

// Pack the account into an Any
accAny, err := codectypes.NewAnyWithValue(myAccount)
if err != nil {
  return nil, err
}

// Create a new Profile with the any.
profile := Profile {
  Account: accAny,
  Bio: "some bio",
}

// We can then marshal the profile as usual.
bz, err := cdc.MarshalBinaryBare(profile)
jsonBz, err := cdc.MarshalJSON(profile)
```

> To summarize, if you want to encode an interface, you need to 1/ pack it into an `Any` and 2/ marshal the `Any`. For convenience, the SDK provides a `MarshalInterface` method to bundle the 2 steps into one. Have a look at [a real-life example in the x/auth module](https://github.com/cosmos/cosmos-sdk/blob/v0.42.1/x/auth/keeper/keeper.go#L218-L221).
> 
> The reverse operation of retrieving the concrete Go type from inside an `Any`, called "unpacking", is done with the `GetCachedValue()` on `Any`.

Reviewer: ```suggestion
To summarize, to encode an interface, you must 1/ pack the interface into an `Any` and 2/ marshal the `Any`. For convenience, the SDK provides a `MarshalInterface` method to bundle these two steps. Have a look at [a real-life example in the x/auth module](https://github.com/cosmos/cosmos-sdk/blob/v0.42.1/x/auth/keeper/keeper.go#L218-L221).

The reverse operation of retrieving the concrete Go type from inside an `Any`, called "unpacking", is done with the `GetCachedValue()` on `Any`.
```

  if p.Account != nil {
    var account AccountI
    return unpacker.UnpackAny(p.Account, &account)
  }

  return nil
}
```

The `UnpackInterfaces` gets called recursively on all structs implementing this method, to allow all `Any`s to have their `GetCachedValue()` correctly populated.

For more information about interface encoding, and especially on `UnpackInterfaces` and how the `Any`'s `type_url` gets resolved using the `InterfaceRegistry`, please refer to [ADR-019](../architecture/adr-019-protobuf-state-encoding.md).

#### `Any` Encoding in the SDK

The above `Profile` example is a fictive example used for educational purposes. In the SDK, we use `Any` encoding in several places (non-exhaustive list):

- the `cryptotypes.PubKey` interface for encoding different types of public keys,
- the `sdk.Msg` interface for encoding different `Msg`s in a transaction,
- the `AccountI` interface for encodinig different types of accounts (similar to the above example) in the x/auth query responses,
- the `Evidencei` interface for encoding different types of evidences in the x/evidence module,
- the `AuthorizationI` interface for encoding different types of x/authz authorizations.

> A real-life example of encoding the pubkey as `Any` inside the Validator struct in x/staking is given below:

Reviewer: ```suggestion
A real-life example of encoding the pubkey as `Any` inside the Validator struct in x/staking is shown in the following example:
```
